crates/bevy_remote/src/lib.rs

```diff
@@ -0,0 +1,961 @@
+}
+
+/// The response to a `GET` request.
+#[derive(Serialize, Deserialize, Clone)]
+struct BrpGetResponse {
+    /// The ID of the entity for which components were requested.
+    entity: Entity,
+
+    /// The values of the requested components.
+    components: HashMap<String, Value>,
+}
+
+/// The response to a `LIST` request.
+#[derive(Serialize, Deserialize, Clone)]
+struct BrpListResponse {
+    /// The ID of the entity for which component names were requested, if
+    /// present.
+    ///
+    /// If this is `None`, then `components` contains the name of all
+    /// reflectable components known to the system.
+    entity: Option<Entity>,
+
+    /// The full type names of the registered components.
+    components: Vec<String>,
+}
+
+/// The response to a `QUERY` request.
+#[derive(Serialize, Deserialize, Clone)]
+struct BrpQueryResponse {
+    /// All results of the query: the entities and the requested components.
+    rows: Vec<BrpQueryRow>,
+}
+
+/// One query match result: a single entity paired with the requested components.
+#[derive(Serialize, Deserialize, Clone)]
+pub struct BrpQueryRow {
+    /// The ID of the entity that matched.
+    pub entity: Entity,
+
+    /// The serialized values of the requested components.
+    #[serde(flatten)]
+    pub components: HashMap<String, Value>,
+}
+
+/// A message from the Bevy Remote Protocol server thread to the main world.
+///
+/// This is placed in the [`BrpMailbox`].
+#[derive(Clone)]
+pub struct BrpMessage {
+    /// The deserialized request from the client.
+    request: BrpRequest,
+
+    /// The channel on which the response is to be sent.
+    ///
+    /// The value sent here is serialized and sent back to the client.
+    sender: Arc<Mutex<Option<OneshotSender<AnyhowResult<Value>>>>>,
+}
+
+/// A resource that receives messages sent by Bevy Remote Protocol clients.
+///
+/// Every frame, the `process_remote_requests` system drains this mailbox, and
+/// processes the messages within.
+#[derive(Resource, Deref, DerefMut)]
+pub struct BrpMailbox(BroadcastReceiver<BrpMessage>);
+
+impl Default for RemotePlugin {
+    fn default() -> Self {
+        RemotePlugin { port: DEFAULT_PORT }
+    }
+}
+
+impl Default for RemoteVerbs {
+    fn default() -> Self {
+        let mut remote_verbs = RemoteVerbs(default());
+        remote_verbs.insert("GET".to_owned(), process_remote_get_request);
+        remote_verbs.insert("QUERY".to_owned(), process_remote_query_request);
+        remote_verbs.insert("SPAWN".to_owned(), process_remote_spawn_request);
+        remote_verbs.insert("INSERT".to_owned(), process_remote_insert_request);
+        remote_verbs.insert("REMOVE".to_owned(), process_remote_remove_request);
+        remote_verbs.insert("DESTROY".to_owned(), process_remote_destroy_request);
+        remote_verbs.insert("REPARENT".to_owned(), process_remote_reparent_request);
+        remote_verbs.insert("LIST".to_owned(), process_remote_list_request);
+        remote_verbs
+    }
+}
```

Comment:
I would like to propose lower casing/snake_casing these instead to be consistent with rust naming conventions for functions, and I also somewhat object to calling these verbs like in REST since they look more like specific RPC calls than general-purpose HTTP verbs.

It might also be worthwhile to be make the names more explicit e.g. "insert_components" instead of "insert" to keep the door open to inserting other things, like maybe a resource.

Reply:
Yeah, I was figuring that resources would be something like `GETRES`/`INSERTRES`/`DESTROYRES`/`LISTRES` and assets would be something like `GETASSET`/`INSERTASSET`/`LISTASSETS`. As with most bikeshedding things like this I have no opinion; to me any convention is as good as any other.

I'd like @coreh to weigh in before I make any changes here.